{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JbSu7kM7VH_A"
      },
      "source": [
        "## First we download the neeeded packages and import them"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 56,
      "metadata": {
        "id": "HnJgXAtLjQ6O"
      },
      "outputs": [],
      "source": [
        "%%capture\n",
        "!pip install pydna\n",
        "!pip install teemi\n",
        "!pip install matplotlib\n",
        "!pip install seaborn\n",
        "!pip install Bio\n",
        "!pip install intermine"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "T2G07ggNGv11"
      },
      "source": [
        "### Function that downloads files from internet"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 57,
      "metadata": {
        "id": "1UzSPwyvGul1"
      },
      "outputs": [],
      "source": [
        "import requests\n",
        "import gzip\n",
        "import pandas as pd\n",
        "import io\n",
        "from Bio import SeqIO\n",
        "\n",
        "\n",
        "# Function to download, unzip, and load into DataFrame\n",
        "def download_unzip_load(url,fasta=False):\n",
        "    # Download the gzipped file\n",
        "    response = requests.get(url, stream=True)\n",
        "    if response.status_code == 200:\n",
        "        # Extract the gzip file\n",
        "        with gzip.open(io.BytesIO(response.content)) as f:\n",
        "            file_content = f.read().decode('utf-8')\n",
        "\n",
        "        # Create DataFrame\n",
        "        data = io.StringIO(file_content)\n",
        "        if fasta:\n",
        "          with gzip.open(io.BytesIO(response.content)) as f:\n",
        "              # Parse the FASTA file\n",
        "              file_content = f.read().decode('utf-8')\n",
        "              data = io.StringIO(file_content)\n",
        "          return data\n",
        "        #df = pd.read_table(data,index_col=0)  # Assuming the file is tab-separated, adjust separator if needed\n",
        "\n",
        "        return data\n",
        "    else:\n",
        "        print(\"Failed to download the file\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9gWIljrbA5Sp"
      },
      "source": [
        "\n",
        "### We define DesignAssembly as we want to tweak the max padding length"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 58,
      "metadata": {
        "id": "5e69f744-8c9c-4498-8f02-4eb41d20f4d4"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import os\n",
        "from Bio.Seq import Seq\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "\n",
        "from Bio.pairwise2 import format_alignment\n",
        "from Bio import pairwise2\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "\n",
        "\n",
        "\n",
        "from teemi.learn.plotting import bar_plot\n",
        "bar_plot\n",
        "from teemi.design.fetch_sequences import read_fasta_files\n",
        "\n",
        "\n",
        "#!/usr/bin/env python\n",
        "# MIT License\n",
        "# Copyright (c) 2023, Technical University of Denmark (DTU)\n",
        "#\n",
        "# Permission is hereby granted, free of charge, to any person obtaining a copy\n",
        "# of this software and associated documentation files (the \"Software\"), to deal\n",
        "# in the Software without restriction, including without limitation the rights\n",
        "# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell\n",
        "# copies of the Software, and to permit persons to whom the Software is\n",
        "# furnished to do so, subject to the following conditions:\n",
        "#\n",
        "# The above copyright notice and this permission notice shall be included in all\n",
        "# copies or substantial portions of the Software.\n",
        "\n",
        "\"\"\" This part of the design module is used for making combinatorial libraries from DNA fragments.\"\"\"\n",
        "\n",
        "# standard libraries\n",
        "import itertools\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "# Pydna for the molecular bio\n",
        "from pydna.design import primer_design\n",
        "from pydna.design import assembly_fragments\n",
        "from pydna.assembly import Assembly\n",
        "from pydna.tm import tm_default as _tm_default\n",
        "from pydna.dseqrecord import Dseqrecord\n",
        "\n",
        "# Typing\n",
        "from typing import List, Callable\n",
        "\n",
        "\n",
        "class DesignAssembly:\n",
        "    \"\"\"Make a combinatorial library from DNA fragments.\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    list_of_seqs : List[List[Dseqrecord]]\n",
        "        A list of a constructs of choice.\n",
        "    list_of_pads : List[Dseqrecord]\n",
        "        A nucleotide sequence to be incorporated into the primers (Max is 40 bp)\n",
        "    positions_of_pads : List[int]\n",
        "        the position in the list of seqs where the pad is incorporated (zero indexed)\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    DesignAssembly object\n",
        "        A powerful class and a lot of information can be retrieved.\n",
        "        Such as: showing all the amplicons needed to construct a combinatorial library\n",
        "        with the simple method --> pcr_list_to_dataframe or primer_list_to_dataframe.\n",
        "\n",
        "    \"\"\"\n",
        "    def __init__(\n",
        "        self,\n",
        "        list_of_seqs: List[List[Dseqrecord]],\n",
        "        list_of_pads: List[List[Dseqrecord]],\n",
        "        positions_of_pads: List[int],\n",
        "        target_tm=55.0,\n",
        "        limit=13,\n",
        "        overlap=35,\n",
        "        tm_func: Callable = _tm_default\n",
        "    ):\n",
        "        ###  1.INITIALIZING ##\n",
        "        self.list_of_seqs = list_of_seqs\n",
        "        self.list_of_pads = list_of_pads\n",
        "        self.positions_of_pads = positions_of_pads\n",
        "\n",
        "        # Generate list of names from Dseqrecord objects\n",
        "        self.list_of_names = [[seq.name for seq in seq_list] for seq_list in list_of_seqs]\n",
        "\n",
        "        ### 2. Amplicons, primers, and their temperatures\n",
        "        (\n",
        "            self.list_of_amplicons,\n",
        "            self.list_of_amplicon_primers,\n",
        "            self.list_of_amplicon_primer_temps,\n",
        "        ) = simple_amplicon_maker(\n",
        "            self.list_of_seqs, self.list_of_names, target_tm=target_tm, limit=limit, primer_tm_func=tm_func\n",
        "        )\n",
        "\n",
        "        # Systematic names\n",
        "        self.systematic_names = get_systematic_names(self.list_of_seqs)\n",
        "\n",
        "        ### 3. COMBINATORIAL LISTS\n",
        "        self.combinatorial_list_of_amplicons = get_combinatorial_list(\n",
        "            self.list_of_amplicons\n",
        "        )\n",
        "        self.combinatorial_list_of_names = get_combinatorial_list(self.list_of_names)\n",
        "        self.combinatorial_list_of_primer_tm = get_combinatorial_list(\n",
        "            self.list_of_amplicon_primer_temps\n",
        "        )\n",
        "\n",
        "        # Making the combinations into a list so we can insert PADS later (They are tuples at this stage, and insert doesnt work for tuples)\n",
        "        for i in range(0, len(self.combinatorial_list_of_amplicons)):\n",
        "            self.combinatorial_list_of_amplicons[i] = list(\n",
        "                self.combinatorial_list_of_amplicons[i]\n",
        "            )\n",
        "\n",
        "        #### 4. Adding PADS ###\n",
        "        for i in range(0, len(self.combinatorial_list_of_amplicons)):\n",
        "            for pads, pos in sorted(zip(self.list_of_pads, self.positions_of_pads), key=lambda x: x[1], reverse=True):\n",
        "                    self.combinatorial_list_of_amplicons[i].insert(pos, pads)\n",
        "\n",
        "        ### 5. Assembling and making overlapping primers\n",
        "        self.list_of_assemblies = assembly_maker(\n",
        "            self.combinatorial_list_of_amplicons, overlap=overlap\n",
        "        )\n",
        "\n",
        "        ### 6. GETTING all primers, annotating, adding features\n",
        "        self.primers = get_primers(\n",
        "            self.list_of_assemblies,\n",
        "            self.combinatorial_list_of_names,\n",
        "            self.combinatorial_list_of_primer_tm,\n",
        "        )\n",
        "\n",
        "        ### 7. Getting Unique primers and re-annotating list_assemblies to get right names\n",
        "        self.unique_primers = unique_primers(self.primers, self.list_of_assemblies)\n",
        "\n",
        "        ### 8. Unique amplicons\n",
        "        self.unique_amplicons = unique_amplicons(self.list_of_assemblies)\n",
        "\n",
        "\n",
        "    def show_contigs(self):\n",
        "        \"\"\"Returns a string of the contigs generated by the assembly\"\"\"\n",
        "        for i in range(0, len(self.list_of_assemblies)):\n",
        "            print(\"\\nContig\" + str(self.systematic_names[i]))\n",
        "            for j in range(0, len(self.list_of_assemblies[i])):\n",
        "                print(\n",
        "                    \"Template: \", self.list_of_assemblies[i][j].name[0:15]\n",
        "                )\n",
        "        return\n",
        "\n",
        "    def show_variants_lib_df(self):\n",
        "        \"\"\"Returns a dataframe of all the variants\"\"\"\n",
        "        combinatorial_lib_variants_df = pd.DataFrame(self.combinatorial_list_of_names)\n",
        "        systematic_names = self.systematic_names\n",
        "        combinatorial_lib_variants_df[\"Systematic_name\"] = systematic_names\n",
        "        combinatorial_lib_variants_df[\"Variant\"] = np.arange(\n",
        "            len(combinatorial_lib_variants_df)\n",
        "        )\n",
        "\n",
        "        return combinatorial_lib_variants_df\n",
        "\n",
        "\n",
        "    def primer_list(self):\n",
        "        \"\"\"Return the list of primers\"\"\"\n",
        "        primer_list = []\n",
        "        for primers in self.unique_primers:\n",
        "            primer_list.append(primers)\n",
        "\n",
        "        return primer_list\n",
        "\n",
        "\n",
        "    def primer_list_to_dataframe(self):\n",
        "        \"\"\"Return a pandas dataframe with list of primers.\"\"\"\n",
        "        df = pd.DataFrame(self.unique_primers)\n",
        "        df.columns = [\n",
        "            \"id\",\n",
        "            \"anneals to\",\n",
        "            \"sequence\",\n",
        "            \"annealing temperature\",\n",
        "            \"length\",\n",
        "            \"price(DKK)\",\n",
        "            \"description\",\n",
        "            \"footprint\",\n",
        "            \"len_footprint\",\n",
        "        ]\n",
        "        return df\n",
        "\n",
        "\n",
        "    def pcr_list_to_dataframe(self):\n",
        "        \"\"\"Prints PCR_list into a pandas dataframe\"\"\"\n",
        "        dataframe_list = []\n",
        "        for i in range(0, len(self.unique_amplicons)):\n",
        "            lst = [\n",
        "                \"PCR{number}\".format(number=i + 1),\n",
        "                self.unique_amplicons[i].name,\n",
        "                self.unique_amplicons[i].forward_primer.id,\n",
        "                self.unique_amplicons[i].reverse_primer.id,\n",
        "                self.unique_amplicons[i].forward_primer.features,\n",
        "                self.unique_amplicons[i].reverse_primer.features,\n",
        "            ]\n",
        "            dataframe_list.append(lst)\n",
        "\n",
        "        df = pd.DataFrame(dataframe_list)\n",
        "        df.columns = [\n",
        "            \"pcr_number\",\n",
        "            \"template\",\n",
        "            \"forward_primer\",\n",
        "            \"reverse_primer\",\n",
        "            \"f_tm\",\n",
        "            \"r_tm\",\n",
        "        ]\n",
        "\n",
        "        return df\n",
        "\n",
        "\n",
        "def get_combinatorial_list(input_list):\n",
        "    \"\"\"\n",
        "    Generates all possible combinations from a list of lists.\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    input_list : list of lists\n",
        "        The input list of lists for which all possible combinations are to be generated.\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    combinations : list of tuples\n",
        "        A list of tuples representing all possible combinations of the elements in the input list of lists.\n",
        "\n",
        "    Example\n",
        "    -------\n",
        "    >>> input_list = [[1, 2], ['a', 'b']]\n",
        "    >>> combinations = get_combinatorial_list(input_list)\n",
        "    >>> print(combinations)\n",
        "    [(1, 'a'), (1, 'b'), (2, 'a'), (2, 'b')]\n",
        "    \"\"\"\n",
        "    combinations = list(itertools.product(*input_list))\n",
        "\n",
        "    return combinations\n",
        "\n",
        "\n",
        "def get_systematic_names(parts_list: list) -> list:\n",
        "    \"\"\"Returns a list of list with systematic names i.e [1,1,1], [1,2,1]... etc\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    parts_list: list of list\n",
        "        can have any type within the list[list[any_type]]\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    systematic_names\n",
        "        list of tuples with the systematic names eg. [(1,1,1),(1,2,1)]\n",
        "\n",
        "    \"\"\"\n",
        "    # The number of parts of each fragment\n",
        "    num_parts = [len(l) for l in parts_list]\n",
        "\n",
        "    ### For naming the strains systematically ### basically making a list from the number of parts with indexes\n",
        "    systematic_lists = []\n",
        "    temp_list = []\n",
        "    for parts in num_parts:\n",
        "        for j in range(parts):\n",
        "            temp_list.append(j + 1)\n",
        "        systematic_lists.append(temp_list)\n",
        "        temp_list = []\n",
        "\n",
        "    # Then we use itertools to make the right combinations\n",
        "    systematic_names = list(itertools.product(*systematic_lists))\n",
        "\n",
        "    return systematic_names\n",
        "\n",
        "def simple_amplicon_maker(\n",
        "    list_of_seqs: list, list_of_names: list, target_tm=56.0, limit=13, primer_tm_func =_tm_default\n",
        "):\n",
        "    \"\"\"Creates amplicons, updates their names\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    list_of_seqs : list[list[pydna.dseqrecord.Dseqrecord]]\n",
        "        List of the pydna.dseqrecord import Dseqrecord elements u want to made into amplicons\n",
        "\n",
        "    list_of_names : list[list[str]]\n",
        "        provide names for the sequences since pydna changes their names to amplicon\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    list_of_amplicons : list[pydna.amplicon.Amplicon]\n",
        "        list with the pydna.amplicon.Amplicon objects that have been made\n",
        "\n",
        "    list_of_amplicon_primers : list[list[(pydna.seq.Seq, pydna.seq.Seq)]]\n",
        "        a list of all the generated primers in tuples where index0 = forward primer\n",
        "        and index1=reverse primer. Both are pydna.seq.Seq objects\n",
        "\n",
        "    list_of_amplicon_primer_temps : list[list[(float, float)]]\n",
        "        a list of melting temperatures in tuples where index0 = forward primer melting temp\n",
        "        and index1=reverse primer melting temp.\n",
        "\n",
        "    \"\"\"\n",
        "    # Start by making an empty list\n",
        "    list_of_amplicons = [[] for i in range(len(list_of_seqs))]\n",
        "    list_of_amplicon_primers = [[] for i in range(len(list_of_seqs))]\n",
        "    list_of_amplicon_primer_temps = [[] for i in range(len(list_of_seqs))]\n",
        "\n",
        "    # Then we calculate the primers with the primer_calc\n",
        "    for i in range(0, len(list_of_seqs)):\n",
        "        for j in range(0, len(list_of_seqs[i])):\n",
        "            # Append Amplicons\n",
        "            amplicons = primer_design(\n",
        "                list_of_seqs[i][j],\n",
        "                tm_func=primer_tm_func,\n",
        "                target_tm=target_tm,\n",
        "                limit=limit,\n",
        "            )\n",
        "\n",
        "            # Updating names\n",
        "            amplicons.name = list_of_names[i][j]\n",
        "            list_of_amplicons[i].append(amplicons)\n",
        "\n",
        "            # Save the primers\n",
        "            primers = (amplicons.forward_primer.seq, amplicons.reverse_primer.seq)\n",
        "            list_of_amplicon_primers[i].append(primers)\n",
        "\n",
        "            # Save melting temps\n",
        "            melting_temps = (\n",
        "                primer_tm_func(str(amplicons.forward_primer.seq)),\n",
        "                primer_tm_func(str(amplicons.reverse_primer.seq)),\n",
        "            )\n",
        "            list_of_amplicon_primer_temps[i].append(melting_temps)\n",
        "\n",
        "    return list_of_amplicons, list_of_amplicon_primers, list_of_amplicon_primer_temps\n",
        "\n",
        "\n",
        "def get_primers(\n",
        "    assemblies: list,\n",
        "    names: list,\n",
        "    primer_temps: list,\n",
        "):\n",
        "    \"\"\"Returns a list of ALL primers from the combinatorial library,\n",
        "    updates names and what they anneal to.\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    assemblies : list[list[pydna.amplicon.Amplicon]]\n",
        "    names : list[(str)]\n",
        "    primer_temps : list[(float, float),..)...]\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    primers : list[list[[pydna.primer.Primer, pydna.primer.Primer]]\n",
        "        All primers that have been made for all assemblies\n",
        "    \"\"\"\n",
        "\n",
        "    all_primers = []\n",
        "\n",
        "    for i, assembly in enumerate(assemblies):\n",
        "        assembly_primers = []\n",
        "        for j, amplicon in enumerate(assembly):\n",
        "            # Names\n",
        "            amplicon.name = names[i][j]\n",
        "            # Primers\n",
        "            if j == 0:  # START OF THE ASSEMBLY\n",
        "                amplicon.forward_primer.description = f\"Anneals to {amplicon.name}\"\n",
        "                amplicon.reverse_primer.description = f\"Anneals to {amplicon.name}, overlaps to {assembly[j + 1].name}\"\n",
        "            elif j == len(assembly) - 1:  # THE END OF THE ASSEMBLY\n",
        "                amplicon.forward_primer.description = f\"Anneals to {amplicon.name}, overlaps to {assembly[j - 1].name}\"\n",
        "                amplicon.reverse_primer.description = f\"Anneals to {amplicon.name}\"\n",
        "            else:  # THE rest:\n",
        "                amplicon.forward_primer.description = f\"Anneals to {amplicon.name}, overlaps to {assembly[j - 1].name}\"\n",
        "                amplicon.reverse_primer.description = f\"Anneals to {amplicon.name}, overlaps to {assembly[j + 1].name}\"\n",
        "\n",
        "            # template it anneals to\n",
        "            amplicon.forward_primer.name = amplicon.name\n",
        "            amplicon.reverse_primer.name = amplicon.name\n",
        "\n",
        "            # Primer tm\n",
        "            amplicon.forward_primer.features = round(float(primer_temps[i][j][0]), 2)\n",
        "            amplicon.reverse_primer.features = round(float(primer_temps[i][j][1]), 2)\n",
        "\n",
        "            assembly_primers.append([amplicon.forward_primer, amplicon.reverse_primer])\n",
        "\n",
        "        all_primers.append(assembly_primers)\n",
        "\n",
        "    return all_primers\n",
        "\n",
        "\n",
        "\n",
        "def assembly_maker(combinatorial_list_of_amplicons: list, overlap=35):\n",
        "    \"\"\"Assembles Amplicons with pad and makes new overlapping primers.\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    combinatorial_list_of_amplicons : list[[pydna.amplicon.Amplicon]]\n",
        "        the list of pydna.amplicon.Amplicon that you want generate\n",
        "        overlapping primers for.\n",
        "    overlap : int (default set to 35)\n",
        "        How many basepair overlaps\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    List_of_assemblies : list[[pydna.amplicon.Amplicon]]\n",
        "        amplicons that overlaps eachother with the specified overlap value.\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "    list_of_assemblies = []\n",
        "    for i in range(0, len(combinatorial_list_of_amplicons)):\n",
        "        list_of_assemblies.append(\n",
        "            assembly_fragments(combinatorial_list_of_amplicons[i], overlap, maxlink=50)\n",
        "        )\n",
        "\n",
        "    return list_of_assemblies\n",
        "\n",
        "\n",
        "def unique_primers(primers: list, list_of_assemblies):\n",
        "    \"\"\"Finds unique primers from a list of assemblies\n",
        "    Parameters\n",
        "    ----------\n",
        "    primers : list[list[list[pydna.primer.Primer]]]\n",
        "        a list of all the primers made for the combinatorial library\n",
        "\n",
        "    list_of_assemblies: list[[pydna.amplicon.Amplicon]]\n",
        "        used here to update the names of the primers\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    unique_primers : list[list(ID,Anneals_to,Sequence,Annealing_temp,Length,Price(DKK))]\n",
        "        Relevant metrics for the unique primers of the combinatorial library.\n",
        "\n",
        "    \"\"\"\n",
        "\n",
        "    unique_primers = []\n",
        "    unique_sequences = []\n",
        "    primer_info = []\n",
        "    counter = 0\n",
        "\n",
        "    ### CHANGING THE NAMES OF THE PRIMERS\n",
        "    for primer_group in primers:\n",
        "        for primer_pair in primer_group:\n",
        "            for primer in primer_pair:\n",
        "                if primer.seq not in unique_sequences:\n",
        "                    unique_sequences.append(primer.seq)\n",
        "                    unique_primers.append(primer)\n",
        "                    primer.id = f\"P{counter+1:03}\"\n",
        "                    counter += 1\n",
        "                    primer_info.append([\n",
        "                        primer.id,\n",
        "                        primer.name,\n",
        "                        primer.seq,\n",
        "                        primer.features,  # anealing temp\n",
        "                        len(primer.seq),  # length\n",
        "                        len(primer.seq) * 1.8,  # price\n",
        "                        primer.description, # description\n",
        "                        primer.footprint,\n",
        "                        len(primer.footprint)\n",
        "                    ])\n",
        "\n",
        "    ### Updating primer names and removing duplicates\n",
        "    for assembly in list_of_assemblies:\n",
        "        for amplicon in assembly:\n",
        "            for primer in unique_primers:\n",
        "                if amplicon.forward_primer.seq == primer.seq:\n",
        "                    amplicon.forward_primer = primer\n",
        "                elif amplicon.reverse_primer.seq == primer.seq:\n",
        "                    amplicon.reverse_primer = primer\n",
        "\n",
        "    return primer_info\n",
        "\n",
        "\n",
        "def unique_amplicons(list_of_assemblies: list):\n",
        "\n",
        "    \"\"\"Finds Unique amplicons from a list of assemblies\n",
        "    Parameters\n",
        "    ----------\n",
        "    list_of_assemblies: list[[pydna.amplicon.Amplicon]]\n",
        "        list of the combinatorial libarary with overlapping ends\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "        unique_amplicons: list[pydna.amplicon.Amplicon]\n",
        "            returns a list of unique amplicons where relavant metrics\n",
        "            are added to the objects.\n",
        "    \"\"\"\n",
        "    ### Unique amplicons\n",
        "    unique_amplicons = []\n",
        "    for i in range(0, len(list_of_assemblies)):\n",
        "        for j in range(0, len(list_of_assemblies[i])):\n",
        "            if list_of_assemblies[i][j] not in unique_amplicons:\n",
        "                unique_amplicons.append(list_of_assemblies[i][j])\n",
        "\n",
        "    return unique_amplicons\n",
        "\n",
        "\n",
        "def get_assembly_figure(assembly_list, limit=15):\n",
        "    \"\"\"\n",
        "    Generates a figure for the specified assembly in the assembly list.\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    assembly_list : list\n",
        "        The list of assemblies.\n",
        "    limit : int, optional\n",
        "        The limit for the assembly, by default 15.\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    contig\n",
        "        The figure for the specified assembly.\n",
        "    \"\"\"\n",
        "    assembly_obj = Assembly(assembly_list, limit=limit)\n",
        "    contig = assembly_obj.assemble_linear()[0].figure()\n",
        "\n",
        "    return contig\n",
        "\n",
        "\n",
        "def count_unique_parts(predictions_df, max_combinations):\n",
        "    \"\"\"\n",
        "    Iterates through the DataFrame of predictions and saves newly encountered parts.\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    predictions_df : pd.DataFrame\n",
        "        DataFrame containing predictions.\n",
        "\n",
        "    max_combinations : int\n",
        "        The maximum number of combinations to consider.\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    encountered_parts : dict\n",
        "        A dictionary containing the unique parts encountered in 'G8H','pG8H', 'pCPR', 'CPR' columns,\n",
        "        total number of unique combinations encountered in 'Sum of parts' and total predictions\n",
        "        encountered in 'Predictions'.\n",
        "    \"\"\"\n",
        "    # Initialization\n",
        "    encountered_parts = {\n",
        "        \"G8H\": [],\n",
        "        \"pG8H\": [],\n",
        "        \"pCPR\": [],\n",
        "        \"CPR\": [],\n",
        "        \"sum_of_parts\": \"\",\n",
        "        \"prediction_number\": \"\",\n",
        "    }\n",
        "    sum_of_parts = 0\n",
        "    prediction_index = 0\n",
        "    g8h_count = 0\n",
        "    cpr_count = 0\n",
        "    pg8h_count = 0\n",
        "    pcpr_count = 0\n",
        "\n",
        "    # Loop through the predictions and save new parts.\n",
        "    while sum_of_parts < max_combinations:\n",
        "        sum_of_parts = g8h_count * cpr_count * pg8h_count * pcpr_count\n",
        "\n",
        "        encountered_parts[\"sum_of_parts\"] = str(sum_of_parts)\n",
        "        encountered_parts[\"prediction_number\"] = str(prediction_index)\n",
        "\n",
        "        g8h = predictions_df.G8H[prediction_index]\n",
        "        pg8h = predictions_df.pG8H[prediction_index]\n",
        "        cpr = predictions_df.CPR[prediction_index]\n",
        "        pcpr = predictions_df.pCPR[prediction_index]\n",
        "        if g8h not in encountered_parts[\"G8H\"]:\n",
        "            encountered_parts[\"G8H\"].append(g8h)\n",
        "            g8h_count += 1\n",
        "        if pg8h not in encountered_parts[\"pG8H\"]:\n",
        "            encountered_parts[\"pG8H\"].append(pg8h)\n",
        "            pg8h_count += 1\n",
        "        if cpr not in encountered_parts[\"CPR\"]:\n",
        "            encountered_parts[\"CPR\"].append(cpr)\n",
        "            cpr_count += 1\n",
        "        if pcpr not in encountered_parts[\"pCPR\"]:\n",
        "            encountered_parts[\"pCPR\"].append(pcpr)\n",
        "            pcpr_count += 1\n",
        "        prediction_index += 1\n",
        "\n",
        "    return encountered_parts\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ab9b438e-494c-430c-baef-05738728bc1a"
      },
      "source": [
        "# 01_1_DESIGN_Promoters"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github"
      },
      "source": [
        "<a target=\"_blank\" href=\"https://colab.research.google.com/github/hiyama341/teemi/blob/main/colab_notebooks/01_1_DESIGN_Promoters.ipynb\">\n",
        "  <img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bb396fed-481c-49ce-b246-6f9aa91e3fbe"
      },
      "source": [
        "## 1. Intro"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ce9ad316-a817-42e1-ae86-10170747e9e4"
      },
      "source": [
        "In this notebook we want to get the seqeunces of the promoters and 5' UTR from the best expressed genes in Chladymonas Reinhardtii based on a RNA seq dataset.\n",
        "We only score the promoters based on average FPKM from 3 different biological repeats of the \"control\" experiments in a salt stress experiment\n",
        "\n",
        "**We want to automate and standardize the selection process with an algorithm**\n",
        "\n",
        "**About the Data used here**:\n",
        "We use the data from a RNA seq experiment with 3 biological replicates of a control with optimal growth conditions (which is the senario we want to mimic since we want to maximize expression). We go with a \"promotor\" definition of 1 kb upstream that is analagous to cerevisae and is in correspondance with other works in Chladymonas reinhardtii: (Scranton et. al 2016 https://www.sciencedirect.com/science/article/pii/S2211926416300443)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "25c0d065-146c-4a1e-8634-2e8f5042f520"
      },
      "source": [
        "### 1.1 **The algorithm/method for finding the set of promoters we're interesed is described as follows:**\n",
        "\n",
        "1. Load in RNA-seq data\n",
        "2. Pre-process the data i.e sort, remove irrelevant data points(zero counts), and set a lower limit for what is considered low-expression, etc.\n",
        "3. Consider what expression dynamics are interesting (in this case high FPKM) and search through the pre-processed data\n",
        "4. Save best candidates.\n",
        "5. Retrieve the sequences of these genes promoters (1kb Upstream sequences of TSS) from the genomic sequence .\n",
        "6. Retrieve the 5' UTR by alinging cDNA with CDS sequences\n",
        "7. Align the sequences to avoid homology.\n",
        "8. Chosen promoters should be saved and exported as FASTA (to an external database).\n",
        "9. Design recombinatorial library of promoter-UTR-GFP\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "17ad465d-57bc-4329-bd92-bf727efdf765"
      },
      "source": [
        "## 2. Data-processing - Change the working directory to your file folder!"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "e1263820-7669-4ee1-8e86-46628ef11227"
      },
      "source": [
        "First step in this workflow is to read the RNAseq data into google colab from the link. Make sure to change the working directory"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 59,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "a2803be3-388f-417e-aaea-f9fb282fcf7a",
        "outputId": "649fe9d7-a191-4ecd-bd04-514e18004a6f"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n",
            "/content/drive/My Drive/Colab Notebooks/DNA\n",
            "/content/drive/MyDrive/Colab Notebooks/DNA\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "print(os.getcwd())\n",
        "os.chdir(\"/content/drive/MyDrive/Colab Notebooks/DNA\") # change to your folder with the data!\n",
        "print(os.getcwd())\n",
        "df = pd.read_table(\"GSE191218_FPKM_matrix.txt\", index_col=0)\n",
        "\n",
        "\n",
        "# URL of the gzipped file\n",
        "#file_url = \"https://ftp.ncbi.nlm.nih.gov/geo/series/GSE191nnn/GSE191218/suppl/GSE191218%5FFPKM%5Fmatrix%2Etxt%2Egz\"\n",
        "#df = pd.read_table(download_unzip_load(file_url),index_col=0)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PuFIU_UQ9nk4"
      },
      "source": [
        "###View the loaded data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 60,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 474
        },
        "id": "d39d7dac-6ffb-4181-ac17-3738ad6305e6",
        "outputId": "1ee0f187-7120-4dbd-8a47-8ce7af8c46ee"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-326ec6ba-ac65-4583-83a6-2dd4a6fa1293\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ck_1</th>\n",
              "      <th>ck_2</th>\n",
              "      <th>ck_3</th>\n",
              "      <th>Salt_2h_1</th>\n",
              "      <th>Salt_2h_2</th>\n",
              "      <th>Salt_2h_3</th>\n",
              "      <th>Salt_4h_1</th>\n",
              "      <th>Salt_4h_2</th>\n",
              "      <th>Salt_4h_3</th>\n",
              "      <th>Salt_8h_1</th>\n",
              "      <th>...</th>\n",
              "      <th>Salt_24h_3</th>\n",
              "      <th>Salt_48h_1</th>\n",
              "      <th>Salt_48h_2</th>\n",
              "      <th>Salt_48h_3</th>\n",
              "      <th>Salt_72h_1</th>\n",
              "      <th>Salt_72h_2</th>\n",
              "      <th>Salt_72h_3</th>\n",
              "      <th>Salt_96h_1</th>\n",
              "      <th>Salt_96h_2</th>\n",
              "      <th>Salt_96h_3</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>gene_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_46276</th>\n",
              "      <td>28389.80301</td>\n",
              "      <td>18450.699590</td>\n",
              "      <td>23802.340680</td>\n",
              "      <td>46826.06517</td>\n",
              "      <td>15044.048030</td>\n",
              "      <td>14574.441280</td>\n",
              "      <td>17904.73074</td>\n",
              "      <td>13978.674280</td>\n",
              "      <td>17194.816690</td>\n",
              "      <td>14463.70159</td>\n",
              "      <td>...</td>\n",
              "      <td>26032.05767</td>\n",
              "      <td>16904.45011</td>\n",
              "      <td>14187.76665</td>\n",
              "      <td>18269.27838</td>\n",
              "      <td>43711.02115</td>\n",
              "      <td>16633.28390</td>\n",
              "      <td>14070.18255</td>\n",
              "      <td>15277.78068</td>\n",
              "      <td>12118.186090</td>\n",
              "      <td>16834.96368</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_132716</th>\n",
              "      <td>24198.10649</td>\n",
              "      <td>14975.504660</td>\n",
              "      <td>14638.736500</td>\n",
              "      <td>30271.59525</td>\n",
              "      <td>10388.870530</td>\n",
              "      <td>8965.308599</td>\n",
              "      <td>14227.58937</td>\n",
              "      <td>10726.211900</td>\n",
              "      <td>15604.009260</td>\n",
              "      <td>10857.34036</td>\n",
              "      <td>...</td>\n",
              "      <td>20961.88423</td>\n",
              "      <td>13019.94416</td>\n",
              "      <td>10013.03383</td>\n",
              "      <td>14835.33927</td>\n",
              "      <td>34135.32094</td>\n",
              "      <td>15249.23905</td>\n",
              "      <td>11601.62224</td>\n",
              "      <td>11189.55128</td>\n",
              "      <td>8000.383790</td>\n",
              "      <td>12087.37016</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_195605</th>\n",
              "      <td>22837.14241</td>\n",
              "      <td>14096.769540</td>\n",
              "      <td>12239.570730</td>\n",
              "      <td>28035.74810</td>\n",
              "      <td>10325.394910</td>\n",
              "      <td>10156.866450</td>\n",
              "      <td>13617.78171</td>\n",
              "      <td>10937.648180</td>\n",
              "      <td>13111.685910</td>\n",
              "      <td>10397.11614</td>\n",
              "      <td>...</td>\n",
              "      <td>25182.68677</td>\n",
              "      <td>13187.96188</td>\n",
              "      <td>10552.94479</td>\n",
              "      <td>14859.29771</td>\n",
              "      <td>33109.91433</td>\n",
              "      <td>15886.54266</td>\n",
              "      <td>11057.88416</td>\n",
              "      <td>10305.78539</td>\n",
              "      <td>8047.358819</td>\n",
              "      <td>10814.89506</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_82172</th>\n",
              "      <td>19254.98557</td>\n",
              "      <td>13162.038170</td>\n",
              "      <td>16929.000200</td>\n",
              "      <td>21909.17463</td>\n",
              "      <td>9175.024738</td>\n",
              "      <td>8505.188908</td>\n",
              "      <td>10885.84487</td>\n",
              "      <td>8573.861486</td>\n",
              "      <td>8388.424316</td>\n",
              "      <td>12030.26047</td>\n",
              "      <td>...</td>\n",
              "      <td>18894.65245</td>\n",
              "      <td>13715.86255</td>\n",
              "      <td>12153.35977</td>\n",
              "      <td>17102.14616</td>\n",
              "      <td>27502.72217</td>\n",
              "      <td>15540.31898</td>\n",
              "      <td>13971.43392</td>\n",
              "      <td>11207.37615</td>\n",
              "      <td>10157.293970</td>\n",
              "      <td>13469.22002</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_195584</th>\n",
              "      <td>20395.72326</td>\n",
              "      <td>9515.468759</td>\n",
              "      <td>8388.463632</td>\n",
              "      <td>20573.79041</td>\n",
              "      <td>9420.561810</td>\n",
              "      <td>7379.889915</td>\n",
              "      <td>10659.89361</td>\n",
              "      <td>10013.819550</td>\n",
              "      <td>15835.382290</td>\n",
              "      <td>12421.90028</td>\n",
              "      <td>...</td>\n",
              "      <td>19966.93666</td>\n",
              "      <td>14059.22548</td>\n",
              "      <td>12324.01483</td>\n",
              "      <td>15087.97748</td>\n",
              "      <td>30580.35797</td>\n",
              "      <td>13946.58369</td>\n",
              "      <td>13582.53302</td>\n",
              "      <td>12328.37024</td>\n",
              "      <td>10047.158930</td>\n",
              "      <td>13252.48737</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_154602</th>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>...</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSRNA049446274</th>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>...</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>novel.1566</th>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>...</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_154956</th>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>...</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_146327</th>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>...</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.00000</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>19267 rows × 27 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-326ec6ba-ac65-4583-83a6-2dd4a6fa1293')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-326ec6ba-ac65-4583-83a6-2dd4a6fa1293 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-326ec6ba-ac65-4583-83a6-2dd4a6fa1293');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-33c34607-1611-41d2-b175-a130437dea84\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-33c34607-1611-41d2-b175-a130437dea84')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-33c34607-1611-41d2-b175-a130437dea84 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "                          ck_1          ck_2          ck_3    Salt_2h_1  \\\n",
              "gene_id                                                                   \n",
              "CHLREDRAFT_46276   28389.80301  18450.699590  23802.340680  46826.06517   \n",
              "CHLREDRAFT_132716  24198.10649  14975.504660  14638.736500  30271.59525   \n",
              "CHLREDRAFT_195605  22837.14241  14096.769540  12239.570730  28035.74810   \n",
              "CHLREDRAFT_82172   19254.98557  13162.038170  16929.000200  21909.17463   \n",
              "CHLREDRAFT_195584  20395.72326   9515.468759   8388.463632  20573.79041   \n",
              "...                        ...           ...           ...          ...   \n",
              "CHLREDRAFT_154602      0.00000      0.000000      0.000000      0.00000   \n",
              "ENSRNA049446274        0.00000      0.000000      0.000000      0.00000   \n",
              "novel.1566             0.00000      0.000000      0.000000      0.00000   \n",
              "CHLREDRAFT_154956      0.00000      0.000000      0.000000      0.00000   \n",
              "CHLREDRAFT_146327      0.00000      0.000000      0.000000      0.00000   \n",
              "\n",
              "                      Salt_2h_2     Salt_2h_3    Salt_4h_1     Salt_4h_2  \\\n",
              "gene_id                                                                    \n",
              "CHLREDRAFT_46276   15044.048030  14574.441280  17904.73074  13978.674280   \n",
              "CHLREDRAFT_132716  10388.870530   8965.308599  14227.58937  10726.211900   \n",
              "CHLREDRAFT_195605  10325.394910  10156.866450  13617.78171  10937.648180   \n",
              "CHLREDRAFT_82172    9175.024738   8505.188908  10885.84487   8573.861486   \n",
              "CHLREDRAFT_195584   9420.561810   7379.889915  10659.89361  10013.819550   \n",
              "...                         ...           ...          ...           ...   \n",
              "CHLREDRAFT_154602      0.000000      0.000000      0.00000      0.000000   \n",
              "ENSRNA049446274        0.000000      0.000000      0.00000      0.000000   \n",
              "novel.1566             0.000000      0.000000      0.00000      0.000000   \n",
              "CHLREDRAFT_154956      0.000000      0.000000      0.00000      0.000000   \n",
              "CHLREDRAFT_146327      0.000000      0.000000      0.00000      0.000000   \n",
              "\n",
              "                      Salt_4h_3    Salt_8h_1  ...   Salt_24h_3   Salt_48h_1  \\\n",
              "gene_id                                       ...                             \n",
              "CHLREDRAFT_46276   17194.816690  14463.70159  ...  26032.05767  16904.45011   \n",
              "CHLREDRAFT_132716  15604.009260  10857.34036  ...  20961.88423  13019.94416   \n",
              "CHLREDRAFT_195605  13111.685910  10397.11614  ...  25182.68677  13187.96188   \n",
              "CHLREDRAFT_82172    8388.424316  12030.26047  ...  18894.65245  13715.86255   \n",
              "CHLREDRAFT_195584  15835.382290  12421.90028  ...  19966.93666  14059.22548   \n",
              "...                         ...          ...  ...          ...          ...   \n",
              "CHLREDRAFT_154602      0.000000      0.00000  ...      0.00000      0.00000   \n",
              "ENSRNA049446274        0.000000      0.00000  ...      0.00000      0.00000   \n",
              "novel.1566             0.000000      0.00000  ...      0.00000      0.00000   \n",
              "CHLREDRAFT_154956      0.000000      0.00000  ...      0.00000      0.00000   \n",
              "CHLREDRAFT_146327      0.000000      0.00000  ...      0.00000      0.00000   \n",
              "\n",
              "                    Salt_48h_2   Salt_48h_3   Salt_72h_1   Salt_72h_2  \\\n",
              "gene_id                                                                 \n",
              "CHLREDRAFT_46276   14187.76665  18269.27838  43711.02115  16633.28390   \n",
              "CHLREDRAFT_132716  10013.03383  14835.33927  34135.32094  15249.23905   \n",
              "CHLREDRAFT_195605  10552.94479  14859.29771  33109.91433  15886.54266   \n",
              "CHLREDRAFT_82172   12153.35977  17102.14616  27502.72217  15540.31898   \n",
              "CHLREDRAFT_195584  12324.01483  15087.97748  30580.35797  13946.58369   \n",
              "...                        ...          ...          ...          ...   \n",
              "CHLREDRAFT_154602      0.00000      0.00000      0.00000      0.00000   \n",
              "ENSRNA049446274        0.00000      0.00000      0.00000      0.00000   \n",
              "novel.1566             0.00000      0.00000      0.00000      0.00000   \n",
              "CHLREDRAFT_154956      0.00000      0.00000      0.00000      0.00000   \n",
              "CHLREDRAFT_146327      0.00000      0.00000      0.00000      0.00000   \n",
              "\n",
              "                    Salt_72h_3   Salt_96h_1    Salt_96h_2   Salt_96h_3  \n",
              "gene_id                                                                 \n",
              "CHLREDRAFT_46276   14070.18255  15277.78068  12118.186090  16834.96368  \n",
              "CHLREDRAFT_132716  11601.62224  11189.55128   8000.383790  12087.37016  \n",
              "CHLREDRAFT_195605  11057.88416  10305.78539   8047.358819  10814.89506  \n",
              "CHLREDRAFT_82172   13971.43392  11207.37615  10157.293970  13469.22002  \n",
              "CHLREDRAFT_195584  13582.53302  12328.37024  10047.158930  13252.48737  \n",
              "...                        ...          ...           ...          ...  \n",
              "CHLREDRAFT_154602      0.00000      0.00000      0.000000      0.00000  \n",
              "ENSRNA049446274        0.00000      0.00000      0.000000      0.00000  \n",
              "novel.1566             0.00000      0.00000      0.000000      0.00000  \n",
              "CHLREDRAFT_154956      0.00000      0.00000      0.000000      0.00000  \n",
              "CHLREDRAFT_146327      0.00000      0.00000      0.000000      0.00000  \n",
              "\n",
              "[19267 rows x 27 columns]"
            ]
          },
          "execution_count": 60,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "df"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3d9409fe-f97b-4166-9716-0acbdd857c6c"
      },
      "source": [
        "### Then we can sort all values by control 1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 61,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 474
        },
        "id": "34760ce8-f54e-45ed-9c1a-85ba9b2b744c",
        "outputId": "aa44e712-939a-4884-eaa3-8b62b0147388"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-6303fed4-10be-4bf6-93a5-5e5914e47ddb\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ck_1</th>\n",
              "      <th>ck_2</th>\n",
              "      <th>ck_3</th>\n",
              "      <th>Salt_2h_1</th>\n",
              "      <th>Salt_2h_2</th>\n",
              "      <th>Salt_2h_3</th>\n",
              "      <th>Salt_4h_1</th>\n",
              "      <th>Salt_4h_2</th>\n",
              "      <th>Salt_4h_3</th>\n",
              "      <th>Salt_8h_1</th>\n",
              "      <th>...</th>\n",
              "      <th>Salt_24h_3</th>\n",
              "      <th>Salt_48h_1</th>\n",
              "      <th>Salt_48h_2</th>\n",
              "      <th>Salt_48h_3</th>\n",
              "      <th>Salt_72h_1</th>\n",
              "      <th>Salt_72h_2</th>\n",
              "      <th>Salt_72h_3</th>\n",
              "      <th>Salt_96h_1</th>\n",
              "      <th>Salt_96h_2</th>\n",
              "      <th>Salt_96h_3</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>gene_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_46276</th>\n",
              "      <td>28389.80301</td>\n",
              "      <td>18450.69959</td>\n",
              "      <td>23802.34068</td>\n",
              "      <td>46826.065170</td>\n",
              "      <td>15044.048030</td>\n",
              "      <td>14574.441280</td>\n",
              "      <td>17904.730740</td>\n",
              "      <td>13978.674280</td>\n",
              "      <td>17194.816690</td>\n",
              "      <td>14463.701590</td>\n",
              "      <td>...</td>\n",
              "      <td>26032.057670</td>\n",
              "      <td>16904.450110</td>\n",
              "      <td>14187.766650</td>\n",
              "      <td>18269.278380</td>\n",
              "      <td>43711.021150</td>\n",
              "      <td>16633.283900</td>\n",
              "      <td>14070.182550</td>\n",
              "      <td>15277.780680</td>\n",
              "      <td>12118.186090</td>\n",
              "      <td>16834.963680</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_159282</th>\n",
              "      <td>25523.92532</td>\n",
              "      <td>13920.94883</td>\n",
              "      <td>16896.95725</td>\n",
              "      <td>27465.606770</td>\n",
              "      <td>10072.984370</td>\n",
              "      <td>7938.195881</td>\n",
              "      <td>11567.986440</td>\n",
              "      <td>9273.011446</td>\n",
              "      <td>12117.918610</td>\n",
              "      <td>9837.408357</td>\n",
              "      <td>...</td>\n",
              "      <td>16898.922790</td>\n",
              "      <td>12219.641070</td>\n",
              "      <td>9349.500125</td>\n",
              "      <td>13153.781830</td>\n",
              "      <td>25594.067340</td>\n",
              "      <td>11694.176500</td>\n",
              "      <td>10373.780760</td>\n",
              "      <td>10037.279400</td>\n",
              "      <td>7434.661709</td>\n",
              "      <td>11152.155140</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_132716</th>\n",
              "      <td>24198.10649</td>\n",
              "      <td>14975.50466</td>\n",
              "      <td>14638.73650</td>\n",
              "      <td>30271.595250</td>\n",
              "      <td>10388.870530</td>\n",
              "      <td>8965.308599</td>\n",
              "      <td>14227.589370</td>\n",
              "      <td>10726.211900</td>\n",
              "      <td>15604.009260</td>\n",
              "      <td>10857.340360</td>\n",
              "      <td>...</td>\n",
              "      <td>20961.884230</td>\n",
              "      <td>13019.944160</td>\n",
              "      <td>10013.033830</td>\n",
              "      <td>14835.339270</td>\n",
              "      <td>34135.320940</td>\n",
              "      <td>15249.239050</td>\n",
              "      <td>11601.622240</td>\n",
              "      <td>11189.551280</td>\n",
              "      <td>8000.383790</td>\n",
              "      <td>12087.370160</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_195605</th>\n",
              "      <td>22837.14241</td>\n",
              "      <td>14096.76954</td>\n",
              "      <td>12239.57073</td>\n",
              "      <td>28035.748100</td>\n",
              "      <td>10325.394910</td>\n",
              "      <td>10156.866450</td>\n",
              "      <td>13617.781710</td>\n",
              "      <td>10937.648180</td>\n",
              "      <td>13111.685910</td>\n",
              "      <td>10397.116140</td>\n",
              "      <td>...</td>\n",
              "      <td>25182.686770</td>\n",
              "      <td>13187.961880</td>\n",
              "      <td>10552.944790</td>\n",
              "      <td>14859.297710</td>\n",
              "      <td>33109.914330</td>\n",
              "      <td>15886.542660</td>\n",
              "      <td>11057.884160</td>\n",
              "      <td>10305.785390</td>\n",
              "      <td>8047.358819</td>\n",
              "      <td>10814.895060</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_195600</th>\n",
              "      <td>21041.90480</td>\n",
              "      <td>11340.48799</td>\n",
              "      <td>11493.08035</td>\n",
              "      <td>22256.013620</td>\n",
              "      <td>7857.944812</td>\n",
              "      <td>7173.794443</td>\n",
              "      <td>10147.575420</td>\n",
              "      <td>7627.639901</td>\n",
              "      <td>9653.927993</td>\n",
              "      <td>9003.050701</td>\n",
              "      <td>...</td>\n",
              "      <td>14592.231860</td>\n",
              "      <td>10375.063040</td>\n",
              "      <td>9293.154489</td>\n",
              "      <td>11921.896710</td>\n",
              "      <td>26673.540990</td>\n",
              "      <td>12428.229490</td>\n",
              "      <td>9570.336491</td>\n",
              "      <td>8076.944475</td>\n",
              "      <td>6626.756473</td>\n",
              "      <td>9235.186696</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>novel.4417</th>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.096871</td>\n",
              "      <td>0.297523</td>\n",
              "      <td>0.069778</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.107485</td>\n",
              "      <td>0.061073</td>\n",
              "      <td>0.062107</td>\n",
              "      <td>...</td>\n",
              "      <td>0.879903</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.072654</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_181126</th>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.078278</td>\n",
              "      <td>0.385528</td>\n",
              "      <td>0.098021</td>\n",
              "      <td>0.042419</td>\n",
              "      <td>0.096410</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>...</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>novel.3136</th>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.686817</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>...</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.330929</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.742806</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_148926</th>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.100884</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.049701</td>\n",
              "      <td>0.042118</td>\n",
              "      <td>...</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.026341</td>\n",
              "      <td>0.032833</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.103929</td>\n",
              "      <td>0.260529</td>\n",
              "      <td>0.269274</td>\n",
              "      <td>0.381261</td>\n",
              "      <td>0.292368</td>\n",
              "      <td>0.256432</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CHLREDRAFT_146327</th>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.00000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>...</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>0.000000</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>19267 rows × 27 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6303fed4-10be-4bf6-93a5-5e5914e47ddb')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-6303fed4-10be-4bf6-93a5-5e5914e47ddb button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-6303fed4-10be-4bf6-93a5-5e5914e47ddb');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-60c04ce6-9f84-4128-9a0d-892c60d8e84c\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-60c04ce6-9f84-4128-9a0d-892c60d8e84c')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-60c04ce6-9f84-4128-9a0d-892c60d8e84c button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "                          ck_1         ck_2         ck_3     Salt_2h_1  \\\n",
              "gene_id                                                                  \n",
              "CHLREDRAFT_46276   28389.80301  18450.69959  23802.34068  46826.065170   \n",
              "CHLREDRAFT_159282  25523.92532  13920.94883  16896.95725  27465.606770   \n",
              "CHLREDRAFT_132716  24198.10649  14975.50466  14638.73650  30271.595250   \n",
              "CHLREDRAFT_195605  22837.14241  14096.76954  12239.57073  28035.748100   \n",
              "CHLREDRAFT_195600  21041.90480  11340.48799  11493.08035  22256.013620   \n",
              "...                        ...          ...          ...           ...   \n",
              "novel.4417             0.00000      0.00000      0.00000      0.096871   \n",
              "CHLREDRAFT_181126      0.00000      0.00000      0.00000      0.000000   \n",
              "novel.3136             0.00000      0.00000      0.00000      0.000000   \n",
              "CHLREDRAFT_148926      0.00000      0.00000      0.00000      0.000000   \n",
              "CHLREDRAFT_146327      0.00000      0.00000      0.00000      0.000000   \n",
              "\n",
              "                      Salt_2h_2     Salt_2h_3     Salt_4h_1     Salt_4h_2  \\\n",
              "gene_id                                                                     \n",
              "CHLREDRAFT_46276   15044.048030  14574.441280  17904.730740  13978.674280   \n",
              "CHLREDRAFT_159282  10072.984370   7938.195881  11567.986440   9273.011446   \n",
              "CHLREDRAFT_132716  10388.870530   8965.308599  14227.589370  10726.211900   \n",
              "CHLREDRAFT_195605  10325.394910  10156.866450  13617.781710  10937.648180   \n",
              "CHLREDRAFT_195600   7857.944812   7173.794443  10147.575420   7627.639901   \n",
              "...                         ...           ...           ...           ...   \n",
              "novel.4417             0.297523      0.069778      0.000000      0.107485   \n",
              "CHLREDRAFT_181126      0.078278      0.385528      0.098021      0.042419   \n",
              "novel.3136             0.000000      0.000000      0.000000      0.686817   \n",
              "CHLREDRAFT_148926      0.100884      0.000000      0.000000      0.000000   \n",
              "CHLREDRAFT_146327      0.000000      0.000000      0.000000      0.000000   \n",
              "\n",
              "                      Salt_4h_3     Salt_8h_1  ...    Salt_24h_3  \\\n",
              "gene_id                                        ...                 \n",
              "CHLREDRAFT_46276   17194.816690  14463.701590  ...  26032.057670   \n",
              "CHLREDRAFT_159282  12117.918610   9837.408357  ...  16898.922790   \n",
              "CHLREDRAFT_132716  15604.009260  10857.340360  ...  20961.884230   \n",
              "CHLREDRAFT_195605  13111.685910  10397.116140  ...  25182.686770   \n",
              "CHLREDRAFT_195600   9653.927993   9003.050701  ...  14592.231860   \n",
              "...                         ...           ...  ...           ...   \n",
              "novel.4417             0.061073      0.062107  ...      0.879903   \n",
              "CHLREDRAFT_181126      0.096410      0.000000  ...      0.000000   \n",
              "novel.3136             0.000000      0.000000  ...      0.000000   \n",
              "CHLREDRAFT_148926      0.049701      0.042118  ...      0.000000   \n",
              "CHLREDRAFT_146327      0.000000      0.000000  ...      0.000000   \n",
              "\n",
              "                     Salt_48h_1    Salt_48h_2    Salt_48h_3    Salt_72h_1  \\\n",
              "gene_id                                                                     \n",
              "CHLREDRAFT_46276   16904.450110  14187.766650  18269.278380  43711.021150   \n",
              "CHLREDRAFT_159282  12219.641070   9349.500125  13153.781830  25594.067340   \n",
              "CHLREDRAFT_132716  13019.944160  10013.033830  14835.339270  34135.320940   \n",
              "CHLREDRAFT_195605  13187.961880  10552.944790  14859.297710  33109.914330   \n",
              "CHLREDRAFT_195600  10375.063040   9293.154489  11921.896710  26673.540990   \n",
              "...                         ...           ...           ...           ...   \n",
              "novel.4417             0.000000      0.000000      0.072654      0.000000   \n",
              "CHLREDRAFT_181126      0.000000      0.000000      0.000000      0.000000   \n",
              "novel.3136             0.330929      0.000000      0.742806      0.000000   \n",
              "CHLREDRAFT_148926      0.026341      0.032833      0.000000      0.103929   \n",
              "CHLREDRAFT_146327      0.000000      0.000000      0.000000      0.000000   \n",
              "\n",
              "                     Salt_72h_2    Salt_72h_3    Salt_96h_1    Salt_96h_2  \\\n",
              "gene_id                                                                     \n",
              "CHLREDRAFT_46276   16633.283900  14070.182550  15277.780680  12118.186090   \n",
              "CHLREDRAFT_159282  11694.176500  10373.780760  10037.279400   7434.661709   \n",
              "CHLREDRAFT_132716  15249.239050  11601.622240  11189.551280   8000.383790   \n",
              "CHLREDRAFT_195605  15886.542660  11057.884160  10305.785390   8047.358819   \n",
              "CHLREDRAFT_195600  12428.229490   9570.336491   8076.944475   6626.756473   \n",
              "...                         ...           ...           ...           ...   \n",
              "novel.4417             0.000000      0.000000      0.000000      0.000000   \n",
              "CHLREDRAFT_181126      0.000000      0.000000      0.000000      0.000000   \n",
              "novel.3136             0.000000      0.000000      0.000000      0.000000   \n",
              "CHLREDRAFT_148926      0.260529      0.269274      0.381261      0.292368   \n",
              "CHLREDRAFT_146327      0.000000      0.000000      0.000000      0.000000   \n",
              "\n",
              "                     Salt_96h_3  \n",
              "gene_id                          \n",
              "CHLREDRAFT_46276   16834.963680  \n",
              "CHLREDRAFT_159282  11152.155140  \n",
              "CHLREDRAFT_132716  12087.370160  \n",
              "CHLREDRAFT_195605  10814.895060  \n",
              "CHLREDRAFT_195600   9235.186696  \n",
              "...                         ...  \n",
              "novel.4417             0.000000  \n",
              "CHLREDRAFT_181126      0.000000  \n",
              "novel.3136             0.000000  \n",
              "CHLREDRAFT_148926      0.256432  \n",
              "CHLREDRAFT_146327      0.000000  \n",
              "\n",
              "[19267 rows x 27 columns]"
            ]
          },
          "execution_count": 61,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "df.sort_values(by=['ck_1'], ascending = False)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "69f0dca6-cf28-4265-bb56-4937e46ad083"
      },
      "source": [
        "It turns out we have a many of genes that do not have any counts in the dataset. We are not interested in these genes since they are inducible or low expression promoters so we filter them out."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1c5ed896-cf80-4bed-afa8-e7651a528605"
      },
      "source": [
        "### 2.1 Removing zeros from the dataset"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 62,
      "metadata": {
        "id": "ecf2e032-cc62-4e23-bd61-1b5c59327923"
      },
      "outputs": [],
      "source": [
        "#print((df==0).all(axis=1))\n",
        "df = df.loc[(df!=0).any(axis=1)]\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3d38c838-5e5d-480d-9212-897685bbb00b"
      },
      "source": [
        "After filtering all the genes out that were only zero, there are still some genes with close to zero counts but we will deal with that later."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "698a5e2d-e658-4602-9ad7-0ceec749368c"
      },
      "source": [
        "### 2.2 Plotting: Gene-expression profiles of C. reinhardtii"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rGSWzktyjQ6f"
      },
      "outputs": [],
      "source": [
        "#Make x range with the length of the dataframe\n",
        "x = np.arange(len(df))\n",
        "y = df['ck_1'].to_list()\n",
        "bar_plot(x, y,\n",
        "        x_label = 'individual genes',\n",
        "         y_label = 'mRNA counts (control 1)'\n",
        "        )"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "b13e775d-45b9-40b9-900c-4240b431b20a"
      },
      "source": [
        "**Notes:**\n",
        "\n",
        "The plotted data shows us that there are approximately 2500 genes that at some point in during the have RNA counts of >100. Most other genes are found to have < 100 counts.  "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "be9c273a-5cbd-4a37-8bc6-92e415bd3992"
      },
      "source": [
        "Additionally, as we saw before quite a lot of data points have RNA counts close to zero and as we are not interested in promoters that do not express anyhting. They can be filtered out."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zyojWVvCjQ6g"
      },
      "source": [
        "### Setting an expression threshold:"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "724513ba-a64b-4f59-82dc-1357e62c127c"
      },
      "source": [
        "We choose to put a threshold expression level - arbitrarily chosen and can differ from dataset to dataset. Genes with counts less than 300 counts combined in the 3 controls are filtered out.  \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "00bd92d1-763e-4da8-9227-4d2e12e6a6d9"
      },
      "outputs": [],
      "source": [
        "threshold = 300\n",
        "df_with_threshold = df[df['ck_1']+ df['ck_2']+ df['ck_3'] > threshold]\n",
        "# adding an extra column with average of ck_1 to ck_3\n",
        "df_with_threshold['ck_average'] = (df_with_threshold['ck_1']+df_with_threshold['ck_2']+df_with_threshold['ck_3'])/3\n",
        "# adding an extra column with standard deviation of ck_1 to ck_3\n",
        "df_with_threshold['ck_std']= df_with_threshold[['ck_1','ck_2','ck_3']].std(axis=1)\n",
        "\n",
        "df_with_threshold"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "25acbe48-4c66-41e2-8c83-668e004ca920"
      },
      "source": [
        "This step filtered out most of the genes with 1019 remaining promoters."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "64235b6f-50b1-4b1d-b2e9-53b174490f4b"
      },
      "source": [
        "### Sorting the values\n",
        "Sorting the values of the dataframe based on average expression in control:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "c4d66d77-d77c-46a2-a39d-73714f2a4200"
      },
      "outputs": [],
      "source": [
        "df_with_threshold = df_with_threshold.sort_values(by=['ck_average'], ascending = False)\n",
        "df_with_threshold.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "s0LQqWN4UPPj"
      },
      "source": [
        "## Now lets plot the 100 highest average expressing genes with std\n",
        "\n",
        "\n",
        "The standard deviation is also ploted, as we want genes that expres uniformly across cells"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "zJuJuD_dUPPj"
      },
      "outputs": [],
      "source": [
        "#plot mean and standard deviation of ck_1 2 and 3\n",
        "plt.figure(figsize=(30,10))\n",
        "plt.errorbar(df_with_threshold[:150].index,df_with_threshold[:150]['ck_average'],df_with_threshold[:150]['ck_std'], linestyle='None', marker='^',color=\"green\")\n",
        "plt.xticks(rotation=90)\n",
        "plt.xlabel('Genes', fontsize=30)\n",
        "plt.ylabel('mRNA counts',fontsize=30)\n",
        "plt.title('Highest expressing genes', fontsize=40)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qrggBQLUjQ6j"
      },
      "source": [
        "### 2.3 Selecting the top 500 genes by average control expression"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "46698b8c-6155-4513-802f-6071411dee9b"
      },
      "source": [
        "We can pick 500 genes with the highest counts and make them into a list."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "e3b4b6e6-ac0b-439e-a5db-02d03e901485"
      },
      "outputs": [],
      "source": [
        "# Selecting top 500\n",
        "List_of_Top100_High_control_prom = df_with_threshold[:500]\n",
        "\n",
        "# making them into a list\n",
        "List_of_Top100_High_control_promoters = List_of_Top100_High_control_prom.index.values.tolist()\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mF0AdnMGAynd"
      },
      "source": [
        "Names of 500 most expressed genes:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "eeae7f46-8068-4ad4-89be-62793b5bc9a7"
      },
      "outputs": [],
      "source": [
        "print(List_of_Top100_High_control_promoters)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "e1f10c1c-8ddd-4153-802a-4a4d873351ed"
      },
      "source": [
        "## 3. Fetching promoter sequences"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "a95b2bf3-59eb-4db4-b68e-a328b60bd83c"
      },
      "source": [
        "#### 3.1 We can now fetch the 1kb upstream of these promoters:\n",
        "\n",
        "According to (Scranton et. al 2016 https://www.sciencedirect.com/science/article/pii/S2211926416300443) the promoters in C. reinhardtii can be regarded as 1 kb upstream.\n",
        "\n",
        "First we load the genome into a dataframe"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6e60c9b4-362d-4f5f-977c-5596907c7c6f"
      },
      "outputs": [],
      "source": [
        "## This reads the fasta file containing all the maped genes:\n",
        "#download_cds=download_unzip_load('https://ftp.ebi.ac.uk/ensemblgenomes/pub/release-57/plants/fasta/chlamydomonas_reinhardtii/cds/Chlamydomonas_reinhardtii.Chlamydomonas_reinhardtii_v5.5.cds.all.fa.gz',fasta=True)\n",
        "#genemaps=read_fasta_files(download_cds)\n",
        "\n",
        "genemaps=read_fasta_files(\"Chlamydomonas_reinhardtii.v3.1.cds.all.fa\")\n",
        "\n",
        "#the following is just to show the loading\n",
        "print(genemaps[:10])"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_k-UrDCwCFnw"
      },
      "source": [
        "### Loading and Putting the cDNA into DataFrame"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wCbl-uAHUPPq"
      },
      "outputs": [],
      "source": [
        "\n",
        "## This reads the fasta file containing all the maped genes:\n",
        "cdna=read_fasta_files(\"Chlamydomonas_reinhardtii.v3.1.cdna.all.fa\")\n",
        "\n",
        "gene_name_l,sequence_l,mapping_l,mapping_id_l=[],[],[],[]\n",
        "for entry in cdna:\n",
        "    description=entry.description\n",
        "    description=description.split(\" \")\n",
        "    gene_name=description[3].replace(\"gene:\",\"\")\n",
        "    seqeunce=entry.seq\n",
        "    mapping=description[2]\n",
        "    mapping_id=mapping.split(\":\")\n",
        "    mapping_id=mapping_id[2]\n",
        "\n",
        "    gene_name_l.append(gene_name)\n",
        "    sequence_l.append(seqeunce)\n",
        "    mapping_l.append(mapping)\n",
        "    mapping_id_l.append(mapping_id)\n",
        "\n",
        "\n",
        "\n",
        "df_cdna = pd.DataFrame({'gene_name':gene_name_l,'sequence':sequence_l,'mapping':mapping_l,'mapping_id':mapping_id_l})\n",
        "df_cdna.head(\n",
        ")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "58c3f245-e679-4c5d-af01-0fd488ffbc52"
      },
      "source": [
        "## Lets extract the locations of the top promoters using the loaded transcription file\n",
        "\n",
        "First i do some testing on how to extract the genomic mapping"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EoXWwuYojQ6x"
      },
      "outputs": [],
      "source": [
        "print(len(genemaps))\n",
        "# Lets make a dataframe containing gene name, seqence and mapping to geneome\n",
        "#some testing\n",
        "genemaps[0].description\n",
        "genemaps[0].description.split(\" \")\n",
        "mapping=genemaps[0].description.split(\" \")[2]\n",
        "mapping_id=mapping.split(\":\")\n",
        "print(mapping,mapping_id)\n",
        "print(\"The mapping read id is \"+ mapping_id[2] + \"From \"+mapping_id[3]+\"to \"+mapping_id[4] )"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YUikItEfDmP0"
      },
      "source": [
        "The mapping are added to the dataframe"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "75Zr5OOcUPPr"
      },
      "outputs": [],
      "source": [
        "df_genemaps = pd.DataFrame(columns=['gene_name','sequence','mapping','mapping_id','genomic_sequence','1kb upstream','cDNA','5UTR','ck_average','ck_std'])\n",
        "\n",
        "gene_name_l,sequence_l,mapping_l,mapping_id_l=[],[],[],[]\n",
        "for entry in genemaps:\n",
        "    description=entry.description\n",
        "    description=description.split(\" \")\n",
        "    gene_name=description[3].replace(\"gene:\",\"\")\n",
        "    seqeunce=entry.seq\n",
        "    mapping=description[2]\n",
        "    mapping_id=mapping.split(\":\")\n",
        "    mapping_id=mapping_id[2]\n",
        "\n",
        "    gene_name_l.append(gene_name)\n",
        "    sequence_l.append(seqeunce)\n",
        "    mapping_l.append(mapping)\n",
        "    mapping_id_l.append(mapping_id)\n",
        "#fix this\n",
        "df_genemaps=pd.DataFrame({'gene_name':gene_name_l,'sequence':sequence_l,'mapping':mapping_l,'mapping_id':mapping_id_l,'genomic_sequence':[\"-\"]*len(gene_name_l),'1kb upstream':[\"-\"]*len(gene_name_l),'cDNA':[\"-\"]*len(gene_name_l),'5UTR':[\"-\"]*len(gene_name_l),'ck_average':[\"-\"]*len(gene_name_l),'ck_std':[\"-\"]*len(gene_name_l)})\n",
        "\n",
        "\n",
        "df_genemaps"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "eMX0J4WkUPPs"
      },
      "outputs": [],
      "source": [
        "#confirmation of dataset\n",
        "print(type(df_genemaps['sequence'][0]),len(df_genemaps['sequence'][0]))\n",
        "df_genemaps['sequence'][0]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "d554c934-2674-44a6-89a2-0ebd85b9530b",
        "outputId": "8e515860-1993-496d-a16f-fdd241f2099a"
      },
      "source": [
        "## The full genome with its ID is added to a pd dataframe"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "0e4fccce-1212-406f-89e5-8909a559adec"
      },
      "outputs": [],
      "source": [
        "genome=read_fasta_files(\"Chlamydomonas_reinhardtii.v3.1.dna.toplevel.fa\")\n",
        "\n",
        "df_genome=pd.DataFrame(columns=['id','sequence'])\n",
        "id_list=[]\n",
        "sequence_list=[]\n",
        "for index,entry in enumerate(genome):\n",
        "    id=entry.id\n",
        "    sequence=entry.seq\n",
        "    id_list.append(id)\n",
        "    sequence_list.append(sequence)\n",
        "\n",
        "df_genome=pd.DataFrame({'id':id_list,'sequence':sequence_list})\n",
        "df_genome"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "H2eQTHQYUPPt"
      },
      "source": [
        "### Extract the Genomic Sequence corresponding to the CDS mappings"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "y1JiJ-GAUPPt"
      },
      "outputs": [],
      "source": [
        "#The start and stop index are extracted from the mapping colunm\n",
        "df_100_highest_control=df_genemaps[df_genemaps['gene_name'].isin(List_of_Top100_High_control_promoters)]\n",
        "df_100_highest_control['index_start']=df_100_highest_control['mapping'].apply(lambda x: x.split(\":\")[3])\n",
        "df_100_highest_control['index_stop']=df_100_highest_control['mapping'].apply(lambda x: x.split(\":\")[4])\n",
        "df_100_highest_control['strand']=df_100_highest_control['mapping'].apply(lambda x: x.split(\":\")[5])\n",
        "\n",
        "#CHLREDRAFT_140045 is there twice because it has two isoforms thus the the df_100 higest has length 101\n",
        "print(sorted(df_100_highest_control[\"gene_name\"].to_list()))\n",
        "print(sorted(List_of_Top100_High_control_promoters))\n",
        "\n",
        "\n",
        "#Now lets extract the sequence from index_start to index_stop for each gene in df_100_highest_control.\n",
        "#The sequence is in df_genome\n",
        "\n",
        "for index, row in df_100_highest_control.iterrows():\n",
        "\n",
        "    #Add cdna to df\n",
        "    df_100_highest_control.loc[index,'cDNA']=df_cdna[df_cdna['gene_name']==row['gene_name']]['sequence'].iloc[0]\n",
        "\n",
        "    #Adding the ck_average and ck_std to the dataframe\n",
        "    df_100_highest_control.loc[index,'ck_average']=df_with_threshold.loc[row['gene_name'],'ck_average']\n",
        "    df_100_highest_control.loc[index,'ck_std']=df_with_threshold.loc[row['gene_name'],'ck_std']\n",
        "\n",
        "    #Because there are two different strands:\n",
        "    if row['strand']==str(1):\n",
        "        df_100_highest_control.loc[index,'genomic_sequence']=df_genome[df_genome['id']==row['mapping_id']]['sequence'].iloc[0][int(row['index_start'])+1:int(row['index_stop'])+1]\n",
        "        df_100_highest_control.loc[index,'1kb upstream']=df_genome[df_genome['id']==row['mapping_id']]['sequence'].iloc[0][int(row['index_start'])+1-1000:int(row['index_start'])+1]\n",
        "\n",
        "    elif row['strand']==str(-1):\n",
        "        forward_genomic_sequence=df_genome[df_genome['id']==row['mapping_id']]['sequence'].iloc[0][int(row['index_start'])+1:int(row['index_stop'])+1]\n",
        "        forward_1kb_upstream=df_genome[df_genome['id']==row['mapping_id']]['sequence'].iloc[0][int(row['index_stop'])+1:int(row['index_stop'])+1+1000]\n",
        "\n",
        "        #These should be reverse complimented\n",
        "        df_100_highest_control.loc[index,'genomic_sequence']=forward_genomic_sequence.reverse_complement()\n",
        "        df_100_highest_control.loc[index,'1kb upstream']=forward_1kb_upstream.reverse_complement()\n",
        "\n",
        "\n",
        "\n",
        "#Write excel and head\n",
        "df_100_highest_control.to_excel(\"100_highest_control.xlsx\")\n",
        "df_100_highest_control.head()\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BK6Iv3vuUPPt"
      },
      "source": [
        "![\n",
        "](attachment:image.png)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tjudrWvfNrxM"
      },
      "source": [
        "### Write the promoters in fasta file"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ZSKUec0GUPPu"
      },
      "outputs": [],
      "source": [
        "#write a fasta with seq.IO for the promoters\n",
        "df_100_highest_control=df_100_highest_control.sort_values(by=['ck_average'], ascending = False)\n",
        "with open(\"100_highest_control_promoters.fasta\", \"w\") as output_handle:\n",
        "    records=[]\n",
        "    for index,row in df_100_highest_control.iterrows():\n",
        "        if len(row['1kb upstream'])==1000 and not 'N' in row['1kb upstream']:\n",
        "            records.append(SeqRecord(row['1kb upstream'],id=row['gene_name'],description=row['mapping']))\n",
        "    #print(records)\n",
        "    SeqIO.write(records, output_handle, \"fasta\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "f3741b26-9f03-429b-9794-2d5415b70756"
      },
      "source": [
        "## 4. Check for Homology between promoters with multiple sequence alingment"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "40f70029-3297-4223-bde0-6aeee1094784"
      },
      "outputs": [],
      "source": [
        "from Bio.Align import MultipleSeqAlignment\n",
        "High_control_promoters = read_fasta_files(\"100_highest_control_promoters.fasta\")\n",
        "All_promoters =  High_control_promoters\n",
        "for i in All_promoters:\n",
        "    if not len(i.seq)==1000:\n",
        "        print(i.id)\n",
        "\n",
        "\n",
        "align = MultipleSeqAlignment(All_promoters)\n",
        "for record in align:\n",
        "    print(record.seq[0:100] + \" \" + record.id)# print(align.substitutions)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "82fb6a9d-976c-4a4d-b4f7-2b625eb9ca65"
      },
      "source": [
        "With biopython it is possible to align sequences in what is termed a \"DUMMY ALIGNMENT\" - see below"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "28f59d9a-5245-49eb-ba2b-334c93f9927a"
      },
      "outputs": [],
      "source": [
        "from Bio.Align import AlignInfo\n",
        "summary_align = AlignInfo.SummaryInfo(align)\n",
        "gap_consensus = summary_align.gap_consensus()\n",
        "consensus = summary_align.dumb_consensus( threshold=0.7)\n",
        "print(consensus)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "2a372f27-13ed-4cb2-b409-3ab967567cc2"
      },
      "outputs": [],
      "source": [
        "counter = 0\n",
        "for nucleotide in consensus:\n",
        "    if nucleotide != 'X':\n",
        "        counter += 1\n",
        "\n",
        "print('Amount of consensus with a 70% threshold: ', (counter/len(consensus))*100, '%')\n",
        "print('Nucleotides that are the same at a 70% threshold:', counter)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xDkCPkIdjQ61"
      },
      "source": [
        "From these results we dont see any long stretches of homology which intuitively makes sense since that would cause genetic instability"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "de142616-2a6e-4df8-8a53-dfde9be7795b"
      },
      "source": [
        "### No we do pairwise alignment of CDS sequences and the full genetic seqeunce (extracted from the genome mapping)\n",
        "\n",
        "This reveals the exons and introns"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "0c3b5349-007a-455d-8e0d-b669a40591d3"
      },
      "outputs": [],
      "source": [
        "for index, row in df_100_highest_control.iterrows():\n",
        "    seq1,seq2 = row[\"genomic_sequence\"],row[\"sequence\"]\n",
        "    #alignments = pairwise2.align.globalxx(seq1, seq2,one_alignment_only=True)\n",
        "    alignments=pairwise2.align.localms(seq1, seq2, match=2, mismatch=-10, open=-5,extend=-.05,one_alignment_only=True,penalize_end_gaps=(True,False))\n",
        "    for alignment in alignments:\n",
        "        print(row['gene_name'])\n",
        "        print(format_alignment(*alignment))\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gkO72QF5UPPx"
      },
      "source": [
        "### Lets try to get the 5' UTR (another better approach is used below)\n",
        "In this code, we map the first 6 and 3 nucleotides of the CDS to the genomic sequence. But it turns out there is an easier way to get the 5' UTR which will be done below\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3kCurfT5UPPy"
      },
      "outputs": [],
      "source": [
        "df_sorted=df_100_highest_control.sort_values(by=['ck_average'], ascending = False)\n",
        "for index, row in df_100_highest_control.iterrows():\n",
        "    seq_genomic,seq_cds = row[\"genomic_sequence\"],row[\"sequence\"]\n",
        "    if seq_cds[0:5] in seq_genomic:\n",
        "        #find index of seq_cds[0:3] in seq_genomic\n",
        "        index=seq_genomic.find(seq_cds[0:6])\n",
        "\n",
        "        print(row['gene_name'],seq_cds[0:6], \"is in position \" ,index)\n",
        "\n",
        "    else:\n",
        "        print(row['gene_name'],seq_cds[0:6], \"is not in genomic sequence\")\n",
        "    if seq_cds[0:3] in seq_genomic:\n",
        "\n",
        "        index2 =seq_genomic.find(seq_cds[0:3])\n",
        "        print(\"The start codon \" + seq_cds[0:3] +\" is in position \",index2)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Ua3f9AAruVO8"
      },
      "source": [
        "### No we do pairwise alignment of cDNA  and the full genetic seqeunce\n",
        "Here we validate that the start of the mapping is actually the TSS"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "yLSJVYw4uU0I"
      },
      "outputs": [],
      "source": [
        "count=1\n",
        "for index, row in df_100_highest_control.iterrows():\n",
        "    if count==100:\n",
        "      break\n",
        "    seq1,seq2 = row[\"genomic_sequence\"],row[\"cDNA\"]\n",
        "    #alignments = pairwise2.align.globalxx(seq1, seq2,one_alignment_only=True)\n",
        "    alignments=pairwise2.align.localms(seq1, seq2, match=2, mismatch=-10, open=-5,extend=-.05,one_alignment_only=True,penalize_end_gaps=(True,False))\n",
        "    for alignment in alignments:\n",
        "        print(row['gene_name'])\n",
        "        print(format_alignment(*alignment))\n",
        "    count=count+1"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ufm9Cru-kF4h"
      },
      "source": [
        "### Getting 5' UTR by Pairwise alignment of CDS and cDNA\n",
        "In the source of the genomic DNA and CDS of C. reinhardtii we also found a file specifying the cDNA sequences. Here we do an alignment of the CDS to the cDNA to find the 5' UTR. This is added to the DataFrame"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "gly0MQAoUPPy"
      },
      "outputs": [],
      "source": [
        "# In this code i want to do a pairwise alignment between the genomic sequence and the cds sequence.\n",
        "for index, row in df_100_highest_control.iterrows():\n",
        "    seq_cdna,seq_cds = row[\"cDNA\"],row[\"sequence\"]\n",
        "    index_cdna=seq_cdna.find(seq_cds[0:6]) # find the index where the cds maps to the cDNA to find the 5' UTR\n",
        "    df_100_highest_control.loc[index,'5UTR']=row['cDNA'][0:index_cdna] # add the 5' UTR to the dataframe\n",
        "    alignments=pairwise2.align.localms(seq_cdna, seq_cds, match=2, mismatch=-10, open=-5,extend=-.05,one_alignment_only=True,penalize_end_gaps=(False,False))\n",
        "    for alignment in alignments:\n",
        "        print(row['gene_name'])\n",
        "        print(format_alignment(*alignment))\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fbaJX-B25DZT"
      },
      "source": [
        "#Plotting the length of the UTR vs the ck_average"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "JXBZCOkAUPPy"
      },
      "outputs": [],
      "source": [
        "# check the dataframe\n",
        "df_100_highest_control.to_excel(\"100_highest_control.xlsx\")\n",
        "df_100_highest_control.head(10)\n",
        "\n",
        "df_sorted=df_100_highest_control.sort_values(by=['ck_average'], ascending = False)\n",
        "# sort the dataframe by ck average\n",
        "#print(np.arange(0,len(df_100_highest_control.index)),df_100_highest_control.sort_values(by=['ck_average'], ascending = False)[\"5UTR\"].apply(lambda x: len(x)).to_list())\n",
        "plt.scatter(df_sorted['ck_average'].to_list(),df_sorted[\"5UTR\"].apply(lambda x: len(x)).to_list(), c=df_sorted[\"ck_average\"].to_list(), cmap='viridis')\n",
        "plt.xlabel(\"Ck average\",fontsize=15)\n",
        "plt.ylabel(\"Length of 5'UTR\",fontsize=15)\n",
        "plt.title(\"Length of 5' UTR vs average expression\",fontsize=20)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_qmpigGFk20o"
      },
      "source": [
        "## Write a Fasta with the UTRs\n",
        "This is done in two different files because the combinatorial library is designed in two different ways (depending on wether it can be as a padding or needs to be a PCR amplified fragment)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "FwHHl_VtUPPz"
      },
      "outputs": [],
      "source": [
        "# sort df based on ck average\n",
        "df_100_highest_control=df_100_highest_control.sort_values(by=['ck_average'], ascending = False)\n",
        "\n",
        "#write a fasta with seq.IO for the 5' UTR\n",
        "with open(\"100_highest_control_promoters_5UTR.fasta\", \"w\") as output_handle:\n",
        "    seen_genes=[]\n",
        "    records=[]\n",
        "    record_short=[]\n",
        "    counter=0\n",
        "    length_list=[]\n",
        "    for index,row in df_100_highest_control.iterrows():\n",
        "        length_list.append(len(row['5UTR']))\n",
        "\n",
        "        if row['gene_name'] in seen_genes:\n",
        "            print(\"This gene is already in the list\" +row['gene_name'])\n",
        "\n",
        "        else:\n",
        "\n",
        "            if len(record_short)+len(records)<10:#since we only want the top 10 UTRs\n",
        "                if len(row['5UTR'])>45:\n",
        "                    records.append(SeqRecord(row['5UTR'],id=\"UTR_\" + row['gene_name'],description=row['mapping']))\n",
        "                    counter=counter+1\n",
        "                elif len(row['5UTR'])>2:\n",
        "                        record_short.append(SeqRecord(row['5UTR'],id=\"UTR_\" + row['gene_name'],description=row['mapping']))\n",
        "        seen_genes.append(row['gene_name'])\n",
        "\n",
        "    SeqIO.write(records, output_handle, \"fasta\")\n",
        "    print(np.average(df_100_highest_control['5UTR'].apply(lambda x: len(x))))\n",
        "\n",
        "with open(\"100_highest_control_promoters_short5UTR.fasta\", \"w\") as output_handle:\n",
        "    SeqIO.write(record_short, output_handle, \"fasta\")\n",
        "    #print(np.average(df_100_highest_control['5UTR'].apply(lambda x: len(x))))\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LeGUl6dFSEgO"
      },
      "source": [
        "### Lets make a histogram of the lengths of the UTR\n",
        "in good accordance with this (from cyano bacterium - because C. reinhardtii not available)\n",
        "https://www.researchgate.net/figure/Distribution-of-UTR-lengths-Histogram-of-the-A-5-0-UTR_fig6_263165770"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "zkXvzH49R-RH"
      },
      "outputs": [],
      "source": [
        "#Make a histogram of the lengths of the 5' UTRs\n",
        "plt.figure(figsize=(10,10))\n",
        "plt.hist(length_list,bins=80,color='green', facecolor='green', alpha=0.75,edgecolor = \"black\")\n",
        "plt.xticks(np.arange(0,1600,100))\n",
        "plt.xlabel(\"5 UTR length (bp)\",fontsize=15)\n",
        "plt.ylabel(\"Number of top 100 most expressed genes\",fontsize=15)\n",
        "plt.grid(True)\n",
        "plt.title(\"Length of 5' UTRs\", fontsize=20)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uKRT-8DcUPPz"
      },
      "source": [
        "## Designing primers for doing a combinatorial libary using gibson"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "1iJ1P5HSUPPz"
      },
      "outputs": [],
      "source": [
        "n_genes=10\n",
        "\n",
        "## load the 5' UTRs and the promoters\n",
        "\n",
        "path = '/Users/andreasrygaard/Documents/Studie/MSC/Synthetic biology'\n",
        "kb_upstream = read_fasta_files('100_highest_control_promoters.fasta')\n",
        "kb_upstream=kb_upstream[:n_genes]\n",
        "UTRs = read_fasta_files('100_highest_control_promoters_5UTR.fasta')\n",
        "short_UTRs=read_fasta_files('100_highest_control_promoters_short5UTR.fasta')\n",
        "GFP=read_fasta_files('GFP_C_reinhardtii.fasta')\n",
        "GFP[0].id=\"Codon_optimised_GFP\"\n",
        "GFP[0].name=\"Codon_optimised_GFP\"\n",
        "GFP[0].description=\"GFP Codon optimised for reinhardtii\"\n",
        "Uphom=[Dseqrecord('CTTCGGTATACAGGCGTGAGCTTCCCCAAGACTGGCCAGACCGTG')]\n",
        "Downhom=[Dseqrecord('ACGGCAAGAAGTTCGACAGCTCCCGCGACCGTGGCGAGCCCTTCT')]\n",
        "\n",
        "\n",
        "## convert into the format of Dseqrecord\n",
        "\n",
        "kb_upstream= [Dseqrecord(seq) for seq in kb_upstream]\n",
        "UTRs = [Dseqrecord(seq) for seq in UTRs]\n",
        "short_UTRs = [Dseqrecord(seq) for seq in short_UTRs]\n",
        "GFP = [Dseqrecord(GFP[0])]\n",
        "\n",
        "print(\"Number of promoters \"+str(len(kb_upstream))+\" Number of >40 nt UTR: \" +str(len(UTRs))+\" Number of <40 nt UTRs \" + str(len(short_UTRs)))\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Ze0F5hW2WBP5"
      },
      "source": [
        "We both have long and short promoters (the stadard implementation of DesignAssembly can only pad with 40 bp). We tweak this number to include the 42 bp longest UTR."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "doQ_KBPzVpOl"
      },
      "source": [
        "### Combinatorial library short and long UTRs\n",
        "\n",
        "First we do the combinatorial libary with the long UTRs. This is not relevant after we tweaked the DesignAssembly but the code is left here"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Q41rHAvtUPP0"
      },
      "outputs": [],
      "source": [
        "# create a list of the 5' UTRs and the promoters\n",
        "list_of_seqs = [kb_upstream, UTRs,GFP]\n",
        "#list_of_seqs = [kb_upstream[0:10],UTRs[0:10]]\n",
        "pad=[Dseqrecord('')]\n",
        "if len(UTRs)>0:\n",
        "  Prom_UTR_combinatorial_library = DesignAssembly(list_of_seqs ,list_of_pads=pad,positions_of_pads=[1], overlap=35, target_tm =58 )\n",
        "Prom_UTR_combinatorial_library.primer_list_to_dataframe()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "C5h4l2fmUPP1"
      },
      "source": [
        "Now we do the combinatorial library with the short_URTs as pads:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bdi8217GUPP1"
      },
      "outputs": [],
      "source": [
        "list_of_sequences_short = [kb_upstream,GFP]\n",
        "UTR_padding=short_UTRs\n",
        "print(UTR_padding)\n",
        "\n",
        "\n",
        "primer_list_df=pd.DataFrame()\n",
        "pcr_list=pd.DataFrame()\n",
        "variant_list=pd.DataFrame()\n",
        "for padding in UTR_padding:\n",
        "  #print(padding)\n",
        "  Prom_short_UTR_combinatorial_library = DesignAssembly(list_of_sequences_short ,list_of_pads=[Uphom,padding,Downhom],positions_of_pads=[-1,0,1], overlap=35, target_tm = 58)\n",
        "  temp_df=Prom_short_UTR_combinatorial_library.primer_list_to_dataframe() #make temp df for primer list\n",
        "  temp_df[\"UTR_id\"]=[padding.id]*len(temp_df)\n",
        "  primer_list_df=pd.concat([temp_df,primer_list_df])\n",
        "\n",
        "  #make temp df for pcr_list\n",
        "  temp_pcr_list=Prom_short_UTR_combinatorial_library.pcr_list_to_dataframe()\n",
        "  temp_pcr_list[\"UTR_id\"]=[padding.id]*len(temp_pcr_list)\n",
        "  pcr_list=pd.concat([temp_pcr_list,pcr_list])\n",
        "\n",
        "  #make variant list\n",
        "\n",
        "  temp_variant=Prom_short_UTR_combinatorial_library.show_variants_lib_df()\n",
        "  temp_variant[\"UTR_id\"]=[padding.id]*len(temp_variant)\n",
        "  variant_list=pd.concat([temp_variant,variant_list])\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "c5pHnbv3Wdqr"
      },
      "source": [
        "### Now we print information about the primers and PCRs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ZCQpu3Jmgplw"
      },
      "outputs": [],
      "source": [
        "\n",
        "primer_list_df\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "edd8e2e3-7d59-4414-b460-29aafb767196"
      },
      "outputs": [],
      "source": [
        "pcr_list\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ZCTw1NfB8xn6"
      },
      "outputs": [],
      "source": [
        "variant_list"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kmn-DQkf82AS"
      },
      "outputs": [],
      "source": [
        "Prom_short_UTR_combinatorial_library.show_variants_lib_df()"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "collapsed_sections": [
        "gkO72QF5UPPx",
        "Ua3f9AAruVO8",
        "ufm9Cru-kF4h"
      ],
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.5"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
